public/app/plugins/datasource/prometheus/querybuilder/components/MetricsLabelsSection.tsx

```diff
@@ -0,0 +1,259 @@
+import React, { useCallback } from 'react';
+
+import { SelectableValue } from '@grafana/data';
+
+import { PrometheusDatasource } from '../../datasource';
+import { getMetadataString } from '../../language_provider';
+import { promQueryModeller } from '../PromQueryModeller';
+import { regexifyLabelValuesQueryString } from '../shared/parsingUtils';
+import { QueryBuilderLabelFilter } from '../shared/types';
+import { PromVisualQuery } from '../types';
+
+import { LabelFilters } from './LabelFilters';
+import { MetricSelect, PROMETHEUS_QUERY_BUILDER_MAX_RESULTS } from './MetricSelect';
+
+export interface MetricsLabelsSectionProps {
+  query: PromVisualQuery;
+  datasource: PrometheusDatasource;
+  onChange: (update: PromVisualQuery) => void;
+  variableEditor?: boolean;
+  onBlur?: () => void;
+}
+
+export function MetricsLabelsSection({
+  datasource,
+  query,
+  onChange,
+  onBlur,
+  variableEditor,
+}: MetricsLabelsSectionProps) {
+  // fixing the use of 'as' from refactoring
+  // @ts-ignore
+  const onChangeLabels = (labels) => {
+    onChange({ ...query, labels });
+  };
+  /**
+   * Map metric metadata to SelectableValue for Select component and also adds defined template variables to the list.
+   */
+  const withTemplateVariableOptions = useCallback(
+    async (optionsPromise: Promise<SelectableValue[]>): Promise<SelectableValue[]> => {
+      const variables = datasource.getVariables();
+      const options = await optionsPromise;
+      return [
+        ...variables.map((value: string) => ({ label: value, value })),
+        ...options.map((option: SelectableValue) => ({
+          label: option.value,
+          value: option.value,
+          title: option.description,
+        })),
+      ];
+    },
+    [datasource]
+  );
+
+  /**
+   * Function kicked off when user interacts with label in label filters.
+   * Formats a promQL expression and passes that off to helper functions depending on API support
+   * @param forLabel
+   */
+  const onGetLabelNames = async (forLabel: Partial<QueryBuilderLabelFilter>): Promise<SelectableValue[]> => {
+    // If no metric we need to use a different method
+    if (!query.metric) {
+      // Todo add caching but inside language provider!
```

Comment:
@gtk-grafana I see it in this PR. Do we track this todo in somewhere?

Reply:
We can remove the todo, the `prometheusResourceBrowserCache` will cache this